### PR TITLE
PS-9704 [8.0]: Escape all RFC 4515 special chars in LDAP filter substitutions

### PR DESCRIPTION
**Pull request by VarunNagaraju**
https://perconadev.atlassian.net/browse/PS-9704

User-supplied values (user_name, user_dn) substituted into LDAP search
filters via {UA} and {UD} were not properly escaped. Only double-quote
received a partial treatment; parentheses, asterisk, and backslash were
passed through verbatim, producing malformed filters that silently broke
group lookups for users whose DN contained those characters.

Add ldap_filter_escape() which applies the mandatory RFC 4515 §3 \XX
hex encoding for LPAREN (\28), RPAREN (\29), ASTERISK (\2a), and
ESC (\5c). Apply it to all three substitution sites: search_dn() for
user_name, and search_groups() for both {UA} (user_name) and {UD}
(user_dn).

Fix heap-use-after-free when granting external roles during login.
acl_authenticate() was calling grant_role() with the acl_user pointer
from mpvio, which is a copy allocated on the connection's mem_root and
freed when the command ends. grant_role() stores ACL_USER by value in
the role graph, retaining the raw user/host char* pointers. Subsequent
DROP USER walks that graph and dereferences those freed pointers, causing
corrupted reads and Warning 1366 on the role_edges table columns.

Fix: look up the durable ACL cache entry via find_acl_user() and pass
that to grant_role() instead of the mem_root copy. This ensures the
role graph holds pointers that remain valid for the lifetime of the
ACL cache.

MTR test exercises DN containing a double-quote, parentheses, an
asterisk, and a backslash — all expected to authenticate and receive
their group role via the dn_test_group groupOfNames entry.

### CODE REVIEW

File: plugin/auth_ldap/src/connection.cc [hunk 1]
```diff
@@ -18,10 +18,43 @@
...
 
 #include "plugin/auth_ldap/include/plugin_log.h"
 
 namespace {
+
+// RFC 4515 §3 requires these five octets to be escaped as \XX
+// when they appear in an LDAP filter assertion value:
+//   NUL (0x00), LPAREN (0x28), RPAREN (0x29), ASTERISK (0x2a), ESC (0x5c)
+std::string ldap_filter_escape(const std::string &input) {
+  std::ostringstream out;
```
Comment (percona-ysorokin) [inline plugin/auth_ldap/src/connection.cc]: Why do we need `std::ostringstream` here? May be just simple std::string::append() / operator +=()

File: plugin/auth_ldap/src/connection.cc [hunk 2]
```diff
@@ -18,10 +18,43 @@
...
+  out << std::hex;
+  for (unsigned char c : input) {
+    switch (c) {
+      case '\0':
```
Comment (jankowsk) [inline plugin/auth_ldap/src/connection.cc]: MySQL doesn't allow '\0', so you may remove this case, see https://dev.mysql.com/doc/refman/8.4/en/identifiers.html
If I'm wrong please add a test for zero.

File: plugin/auth_ldap/src/connection.cc [hunk 3]
```diff
@@ -18,10 +18,43 @@
...
+        out << "\\29";
+        break;
+      case '*':
+        out << "\\2a";
+        break;
+      case '\\':
+        out << "\\5c";
+        break;
+      default:
+        out << c;
```
Comment (percona-ysorokin) [inline plugin/auth_ldap/src/connection.cc]: std::ostream interprets `unsigned char`s as just `char`s (not as unsigned integers) and prints them as a single character.
If the intent here is to print in hex, then
```c++
out << static_cast<std::uint16_t>(c)
```
Also if you need 2 characters for you hex representation, then
```c++
out << std::hex << std::setfill('0');
...
out << std::setw(2) << static_cast<std::uint16_t>(c);
```

File: plugin/auth_ldap/src/connection.cc (hunk 1 above)
Comment (VarunNagaraju) [reply plugin/auth_ldap/src/connection.cc]: Done.

File: plugin/auth_ldap/src/connection.cc (hunk 2 above)
Comment (VarunNagaraju) [reply plugin/auth_ldap/src/connection.cc]: Removed the case for '\0'.

File: plugin/auth_ldap/src/connection.cc (hunk 3 above)
Comment (VarunNagaraju) [reply plugin/auth_ldap/src/connection.cc]: Modified the method to use std:string.

File: sql/auth/sql_authorization.cc
Comment (jankowsk) [inline sql/auth/sql_authorization.cc:1]: Please use commit [PS-11253 [8.4]: Fix heap-use-after-free when granting external roles](https://github.com/percona/percona-server/pull/5941/changes/2ac57c4b5554432bbac4b91c9b18b173eb400b7c) instead this.

Comment (VarunNagaraju) [reply sql/auth/sql_authorization.cc:1]: Done

File: plugin/auth_ldap/src/connection.cc [hunk 4]
```diff
@@ -22,6 +22,37 @@
...
 namespace {
+
+// RFC 4515 §3 requires these five octets to be escaped as \XX
+// when they appear in an LDAP filter assertion value:
+//   NUL (0x00), LPAREN (0x28), RPAREN (0x29), ASTERISK (0x2a), ESC (0x5c)
+std::string ldap_filter_escape(const std::string &input) {
+  std::string out;
+  out.reserve(input.size() * 3);  // worst case every byte becomes "\XX"
+
+  for (unsigned char c : input) {
```
Comment (percona-ysorokin) [inline plugin/auth_ldap/src/connection.cc]: Why do we have `unsigned char` here?

**Review by jankowsk — COMMENTED** [review]
I'd recommend having sql/auth/sql_authentication.cc fix in a separate commit as it is not directly linked to this PS. You may use a proper description of the problem and link to PS in the commit message. See https://github.com/percona/percona-server/pull/5941/changes/60a906a6b39ae44b955e17fb3c7eeefe89e988fc

File: plugin/auth_ldap/src/connection.cc (hunk 4 above)
Comment (VarunNagaraju) [reply plugin/auth_ldap/src/connection.cc]: Typically, raw byte manipulation is done using unsigned char and It covers the cases in the future if the input contains any non-ascii values, then the comparison would be accurate and the default case casts back to char.
But, we could modify it to char since MySQL usernames and LDAP DNs, both are constrained to valid UTF-8.

Comment (percona-ysorokin) [reply plugin/auth_ldap/src/connection.cc]: My point here is that you have a sequence of `char`s (`std::string`), than you iterate over individual characters performing conversions to `unsigned char`s, and after that you use switch `statement` in which you have `case:` clauses with literals of type `char` (`'('` for instance). Moreover, in the `default:` clause you do `static_cast back` to original `char` again. This can work (I do not see any problems that these conversions will cause any value-changing side effects), but I just don't see the reason why we need it as we do not do any computations on the unsigned values here.

Comment (VarunNagaraju) [reply plugin/auth_ldap/src/connection.cc]: Right. The sole reason was to handle any non-ascii values in the future. But, it's not entirely necessary for now. Will update it to char.

**Comment by hulyav** [comment]
Based on the discussion during the team meeting, not needed anymore.
